File: game/level-generator.ts
import { Block } from "./block"
import { Food } from "./food"
import { analyzeMoveAbsolute, dragSnake, takeMove } from "./game-logic"
import { deserialize, entities, levelInfo, serialize, undo } from "./game-state"
import { hitTestAllEntities, invertCollisionLayer, layersCollide, shuffle, topLayer, withinLevel } from "./helpers"
import Snake from "./snake"
import { CollisionLayer, DIRECTIONS, Move } from "./types"

export async function generateLevel() {
  const tries = 20
  let bestComplexity = 0
  let bestLevel = null
  for (let i = 0; i < tries; i++) {
    const complexity = tryGenerateLevel().estimatedPuzzleComplexity
    await new Promise(resolve => setTimeout(resolve, 100)) // give the UI a chance to update, avoid crashing the browser
    if (complexity > bestComplexity) {
      bestComplexity = complexity
      bestLevel = serialize()
    }
  }
  console.log("Best complexity found:", bestComplexity)
  if (bestLevel) {
    deserialize(bestLevel)
  }
}

function tryGenerateLevel() {
  const puzzleGenerationLimit = 10000
  const targetPuzzleComplexity = 100
  const blockDensity = 0.3
  const foodChance = 0.3

  // I think smaller levels should be statistically more likely to generate
  // "puzzle"-like levels rather than meaningless traversal.
  levelInfo.width = 2 + Math.floor(Math.random() * 5)
  levelInfo.height = 2 + Math.floor(Math.random() * 5)
  entities.length = 0

  // Create blocks
  for (let x = 0; x < levelInfo.width; x++) {
    for (let y = 0; y < levelInfo.height; y++) {
      if (Math.random() < blockDensity) {
        const block = new Block()
        block.x = x
        block.y = y
        entities.push(block)
      }
    }
  }

  // Create snakes
  const nSnakes = 1 + Math.floor(Math.random() * 3)
  for (let i = 0; i < nSnakes; i++) {
    const snake = new Snake()
    entities.push(snake) // push early so that hitTestAllEntities includes the snake itself
    snake.segments.length = 0
    let x = Math.floor(Math.random() * levelInfo.width)
    let y = Math.floor(Math.random() * levelInfo.height)
    const hits = hitTestAllEntities(x, y)
    let layer = CollisionLayer.White
    if (hits.length > 0) {
      layer = invertCollisionLayer(hits[0].layer)
    }
    snake.segments.push({ x, y, width: 1, height: 1, layer })
    const targetSnakeEndLength = 2 + Math.floor(Math.random() * 10)
    for (let j = 1; j < targetSnakeEndLength; j++) {
      // Try to place the next segment in a random direction
      const directions = [...DIRECTIONS]
      shuffle(directions)
      for (const direction of directions) {
        if (!withinLevel({ x: x + direction.x, y: y + direction.y, width: 1, height: 1 })) {
          continue
        }
        const hits = hitTestAllEntities(x + direction.x, y + direction.y)
        if (!layersCollide(topLayer(hits), layer)) {
          x += direction.x
          y += direction.y
          snake.segments.push({ x, y, width: 1, height: 1, layer })
          break
        }
      }
    }
  }

  // Simulate in reverse, occasionally creating collectables and shrinking snakes as they move backwards
  const moves: Move[] = []
  for (let i = 0; i < puzzleGenerationLimit; i++) {

    const snakes = entities.filter(e => e instanceof Snake) as Snake[]
    const snake = snakes[Math.floor(Math.random() * snakes.length)]
    const direction = DIRECTIONS[Math.floor(Math.random() * DIRECTIONS.length)]
    const potentialBeforeTile = {
      x: snake.segments[snake.segments.length - 1].x - direction.x,
      y: snake.segments[snake.segments.length - 1].y - direction.y,
      width: 1,
      height: 1,
    }
    if (!withinLevel(potentialBeforeTile)) {
      continue
    }
    // TODO: technically should ignore opposite end of the snake since moving onto your tail is valid
    // I have logic to conditionally ignore the tail, but would need to conditionally ignore the head since we're simulating backwards
    const hits = hitTestAllEntities(potentialBeforeTile.x, potentialBeforeTile.y)
    if (!layersCollide(topLayer(hits), snake.segments[0].layer)) {
      // I originally added undoable() here to debug the level generation,
      // then used it for the core logic of level generation, for backtracking.
      // However, it's inefficient, ESPECIALLY if we're also calling serialize()
      // and I was getting lots of browser tab crashes, so I optimized it to
      // use deserialize() instead, and restore the `growOnNextMove` property
      // which is the only thing changed before the snapshot.
      // Still getting browser tab crashes, though,
      // so it'd be better to avoid serialization altogether.
      // Or reimplement this in C or something.
      // undoable()
      const prevGrowOnNextMove = snake.growOnNextMove
      const eat = Math.random() < foodChance && snake.segments.length > 1
      // `growOnNextMove` is supposed to be set after eating,
      // so we have to do it before the reverse move, and before the `expected` snapshot,
      // because 
      snake.growOnNextMove = eat
      const expected = serialize()
      const previousHead = { ...snake.segments[0] }
      // FIXME: it's not validating in the case that it generates a collectable
      if (eat) {
        const food = new Food()
        food.x = snake.segments[0].x
        food.y = snake.segments[0].y
        food.layer = snake.segments[0].layer
        entities.push(food)
      }
      dragSnake(snake, snake.segments.length - 1, potentialBeforeTile)
      if (eat) {
        snake.segments.pop() // shrink the snake, TODO: is this the right end?
        // snake.segments.shift()
      }
      const move = analyzeMoveAbsolute(snake, previousHead)
      if (!move.valid) {
        // console.log("Undoing generated invalid move:", move)
        // backtrack if the move is invalid
        deserialize(expected); snake.growOnNextMove = prevGrowOnNextMove // undo()
        continue
      }
      // Also need to check that game state matches exactly if simulating forwards
      // because the move may be valid, but it won't give the expected game state.
      // Entities may be ordered differently.
      // Note: this MAY be too limiting, comparing the total entity order
      // Comparing some sort of partial order may be better, but more complex and error-prone.
      // I haven't determined that it's necessary, but this may be subtly rejecting
      // more interesting puzzles, if there's a case where the entities are
      // effectively ordered the same, but irrelevant disorder exists,
      // and this aligns with characteristics of interesting puzzles.
      takeMove(move)
      const actual = serialize()
      undo() // always undo takeMove done just for validation
      if (actual !== expected) {
        // console.log("Undoing generated move which gave an inconsistent game state:", {
        //   expected,
        //   actual,
        //   move,
        // })
        // backtrack if validation failed
        deserialize(expected); snake.growOnNextMove = prevGrowOnNextMove // undo()
        continue
      }
      moves.push(move)
      if (moves.length >= targetPuzzleComplexity) {
        break
      }
    }
  }

  // TODO: for a better complexity estimate, we could try to contract the playthrough
  // by removing moves that are not necessary to solve the puzzle.
  // (We could also crop the level afterwards, to the bounding box of moved snakes across all moves,
  // although it wouldn't affect the complexity.)
  const numFood = entities.filter(e => e instanceof Food).length
  const puzzleSteps = moves.length
  const totalMoveComplexity = moves.reduce((sum, move) => {
    return sum + 1 + (move.entitiesThere.length * 2 + move.entitiesToPush.length * 3)
  }, 0)
  const stats = {
    puzzleSteps,
    numFood,
    totalMoveComplexity,
    // totalMoveComplexity already includes puzzleSteps, numFood, effectively
    estimatedPuzzleComplexity: totalMoveComplexity
  }
  console.log("Generated a puzzle that can be solved in", puzzleSteps, "steps (likely fewer)", stats)
  return stats
}

File: game/go/level-generator.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"slices"
)

func GenerateLevel() (*Level, error) {
	const tries = 200
	var bestComplexity int
	var bestLevel *Level
	for i := 0; i < tries; i++ {
		level, complexity := tryGenerateLevel()
		// time.Sleep(100 * time.Millisecond)
		if level != nil && complexity > bestComplexity {
			bestComplexity = complexity
			bestLevel = level
		}
	}
	fmt.Fprintf(os.Stderr, "Best complexity found: %d\n", bestComplexity)
	if bestLevel == nil {
		return nil, fmt.Errorf("failed to generate any valid level in %d tries", tries)
	}
	return bestLevel, nil
}

func tryGenerateLevel() (*Level, int) {
	const puzzleGenerationLimit = 10000
	const targetPuzzleComplexity = 1000
	const blockDensity = 0.3
	const foodChance = 0.9

	// I think smaller levels should be statistically more likely to generate
	// "puzzle"-like levels rather than meaningless traversal.
	width := rand.Intn(5) + 2
	height := rand.Intn(5) + 2

	level := &Level{
		Info:     LevelInfo{Width: width, Height: height},
		Grid:     make([][]CollisionLayer, height),
		Entities: make([]Entity, 0, 100),
	}

	// Initialize grid with random block types
	for i := range level.Grid {
		level.Grid[i] = make([]CollisionLayer, width)
		for j := range level.Grid[i] {
			if rand.Float32() < blockDensity {
				level.Grid[i][j] = White
			} else {
				level.Grid[i][j] = Black
			}
		}
	}

	// Create snakes
	numSnakes := rand.Intn(3) + 1
	for i := 0; i < numSnakes; i++ {
		x := rand.Intn(width)
		y := rand.Intn(height)
		// Get layer before appending snake so we don't retrieve the snake's own (uninitialized) layer
		layer := invertCollisionLayer(topLayerAt(x, y, level))
		// append early (before topLayerAt) so that hit tests include the snake itself
		level.Entities = append(level.Entities, &Snake{ID: fmt.Sprint(i + 1)})
		snake := level.Entities[i].(*Snake)
		snake.Segments = []Point{{X: x, Y: y}}
		snake.Layer = layer
		targetSnakeEndLength := 2 + rand.Intn(10)
		for j := 1; j < targetSnakeEndLength; j++ {
			// Try to place the next segment in a random direction
			directionOrder := rand.Perm(len(CardinalDirections))
			for _, directionIndex := range directionOrder {
				direction := CardinalDirections[directionIndex]
				if !withinLevel(Point{X: x + direction.X, Y: y + direction.Y}, level) {
					continue
				}
				if !layersCollide(topLayerAt(x+direction.X, y+direction.Y, level), layer) {
					x += direction.X
					y += direction.Y
					snake.Segments = append(snake.Segments, Point{X: x, Y: y})
					break
				}
			}
		}
	}

	// Simulate in reverse, occasionally creating collectables and shrinking snakes as they move backwards
	var moves []Move
	for i := 0; i < puzzleGenerationLimit; i++ {
		snakes := getSnakes(level)
		snake := snakes[rand.Intn(len(snakes))]
		direction := CardinalDirections[rand.Intn(len(CardinalDirections))]
		potentialBeforeTile := Point{
			X: snake.Segments[len(snake.Segments)-1].X - direction.X,
			Y: snake.Segments[len(snake.Segments)-1].Y - direction.Y,
		}
		if !withinLevel(potentialBeforeTile, level) {
			continue
		}
		// TODO: technically should ignore opposite end of the snake since moving onto your tail is valid
		// I have logic to conditionally ignore the tail, but would need to conditionally ignore the head since we're simulating backwards
		hits := hitTestAllEntities(potentialBeforeTile.X, potentialBeforeTile.Y, level, HitTestOptions{
			// IgnoreHeadOfSnake: snake,
		})
		if !layersCollide(topLayer(hits), snake.Layer) {
			prevGrowOnNextMove := snake.GrowOnNextMove
			// const eat = Math.random() < foodChance && snake.segments.length > 1
			previousHead := snake.Segments[0]
			eat := rand.Float32() < foodChance && len(snake.Segments) > 1 &&
				// prevent generating food on top of other food
				!slices.ContainsFunc(hitTestAllEntities(previousHead.X, previousHead.Y, level, HitTestOptions{}), func(hit Hit) bool {
					_, isFood := hit.Entity.(*Food)
					return isFood
				})
			// `GrowOnNextMove` is supposed to be set after eating,
			// so we have to do it before the reverse move, and before the `expected` snapshot
			snake.GrowOnNextMove = eat
			expected := copyLevel(level)
			// FIXME: it's not validating in the case that it generates a collectable
			if eat {
				food := &Food{}
				food.Position = previousHead
				food.Layer = snake.Layer
				level.Entities = append(level.Entities, food)
			}
			// dragSnake(snake, len(snake.Segments)-1, potentialBeforeTile)
			moveSnakeByTail(snake, potentialBeforeTile)
			if eat {
				snake.Segments = snake.Segments[:len(snake.Segments)-1] // shrink the snake, TODO: is this the right end?
			}
			move := AnalyzeMoveAbsolute(snake, previousHead, level)
			if !move.Valid {
				// fmt.Fprintf(os.Stderr, "Undoing generated invalid move:", move)
				// backtrack if the move is invalid
				level = expected
				snake.GrowOnNextMove = prevGrowOnNextMove
				continue
			}
			// Also need to check that game state matches exactly if simulating forwards
			// because the move may be valid in isolation, but not as a way to get to the expected state.
			// In other words, it can be valid move without being a valid precondition.
			// Entities may be ordered differently.
			// Note: this MAY be too limiting, comparing the total entity order
			// Comparing some sort of partial order may be better, but more complex and error-prone.
			// I haven't determined that it's necessary, but this may be subtly rejecting
			// more interesting puzzles, if there's a case where the entities are
			// effectively ordered the same, but irrelevant disorder exists,
			// and this aligns with characteristics of interesting puzzles.
			beforeMove := copyLevel(level)
			TakeMove(move, level)
			actual := copyLevel(level)
			level = beforeMove // always undo takeMove done just for validation
			// if actual != expected {
			if !Equal(expected, actual) {
				// expectedJSON, _ := SerializeLevel(expected)
				// actualJSON, _ := SerializeLevel(actual)
				// fmt.Fprint(os.Stderr, "Undoing generated move which gave an inconsistent game state:", map[string]interface{}{
				// 	"expectedJSON":               string(expectedJSON),
				// 	"actualJSON":                 string(actualJSON),
				// 	"expectedJSON == actualJSON": slices.Equal(expectedJSON, actualJSON),
				// 	"move":                       move,
				// })

				// // Save expected and actual to files for debugging
				// os.WriteFile("expected.json", expectedJSON, 0644)
				// os.WriteFile("actual.json", actualJSON, 0644)
				// os.Exit(1)

				// backtrack if validation failed
				level = expected
				snake.GrowOnNextMove = prevGrowOnNextMove
				continue
			}
			moves = append(moves, move)
			if len(moves) >= targetPuzzleComplexity {
				break
			}
		}
	}

	// It's easy to think of metrics that, if applied to human-designed levels,
	// would correlate with puzzle complexity, but it's hard to design rules
	// that filter out garbage generated levels...
	// For instance, more snake switches might imply a more complex puzzle solution,
	// but in practice it favors open-ended levels with more snakes than you need.
	// Some things we might try:
	// - Automatically simplify the playthrough by removing unnecessary moves.
	//   - Might be difficult because removing one move at a time can make it fail to solve
	//     even if the move is part of a sequence of useless moves.
	//   - Could look for any duplicate level states in the playthrough,
	//     and remove anything between the first and last occurrence of that state
	//     (and one of the occurrences).
	//   - Before looking for cycles (i.e. duplicates),
	//     we could try to sort the moves by snake,
	//     in as much as it preserves the validity of the playthrough.
	//     This would help prevent a disorganized playthrough from inflating the complexity score.
	//   - We could also try to generatively simplify the playthrough.
	//     It may be too expensive to find a playthrough from scratch,
	//     trying every move in every state, but since we have a valid playthrough,
	//     we could try every combination of N moves at each step, and see if it
	//     results in a later state in the playthrough.
	//     If it does, and it's shorter than the original subsequence,
	//     we could replace the original subsequence with the shorter one.
	// - Count the number of moves where the possible moves at that state are limited.
	//   - This may incentivize tighter, less open-ended levels.
	//   - In the extreme, this would favor a level that is just a corridor.
	//     However, since we're currently generating too open-ended levels,
	//		 it may help to balance out the structure.
	//   - We could call these "pinch points" in the puzzle,
	//     but again, if we're measuring it in not the puzzle or an optimal playthrough,
	//     but rather some stupid random thoughtless playthrough,
	//     this may incentivize meaningless "wall hugging" if we can't
	//     figure out how to optimize the playthrough well enough.
	// - Count the times a snake moves onto another snake WHICH IS USED LATER TO EAT FOOD.
	//   - Without checking that the snake is used later, this would favor
	//     levels that unnecessarily/meaninglessly use snakes as terrain.
	//     Using snakes as terrain is core to the game, but only as a means to an end;
	//     by itself it only increases _visual_ complexity.
	//     It might make it harder to solve, but not in the right way.
	//     We don't just want red herrings or visual noise,
	//     we want sequences where you have to use snakes together.
	//   - May or may not want to count only when A moves onto B when A was not already on B.
	//     Or compromise by scoring both cases differently. I dunno.

	lastSnakeID := ""
	numSnakeSwitches := 0
	for _, move := range moves {
		// complexity += 1 + len(move.EntitiesThere)*2 //+ len(move.entitiesToPush) * 3
		// complexity += len(move.EntitiesThere)
		if move.Snake.ID != lastSnakeID {
			numSnakeSwitches++
		}
		lastSnakeID = move.Snake.ID
	}
	numFood := 0
	for _, entity := range level.Entities {
		if _, ok := entity.(*Food); ok {
			numFood++
		}
	}
	complexity := numFood //+ numSnakeSwitches
	if numFood == 0 {
		// Unsolvable
		return nil, 0
	}
	return level, complexity
}
